### PR TITLE
resource/server: support boot_type

### PR DESCRIPTION
**Pull request by nicolai86**
this PR adds support for the [newly introduced boot_type](https://blog.online.net/2018/04/19/scaleway-feature-improvements-custom-kernel-security-group-live-reload-imagehub-packer/) attribute for scaleway server resources:

```
resource "scaleway_server" "base" {
  name = "test"
  image = "e20532c4-1fa0-4c97-992f-436b8d372c07"
  type = "VC1S"
  tags = [ "terraform-test" ]
  boot_type = "local"
}
```

valid values are `local` and `bootscript` for x86_64, as well as `bootscript` for arm instance types.

Tests are green:

```
=== RUN   TestAccScalewayDataSourceBootscript_Filtered
--- PASS: TestAccScalewayDataSourceBootscript_Filtered (8.64s)
=== RUN   TestAccScalewayDataSourceImage_Basic
--- PASS: TestAccScalewayDataSourceImage_Basic (26.29s)
=== RUN   TestAccScalewayDataSourceImage_Filtered
--- PASS: TestAccScalewayDataSourceImage_Filtered (23.28s)
=== RUN   TestValidateBootType
=== RUN   TestValidateBootType/local
=== RUN   TestValidateBootType/bootscript
=== RUN   TestValidateBootType/remote
--- PASS: TestValidateBootType (0.00s)
    --- PASS: TestValidateBootType/local (0.00s)
    --- PASS: TestValidateBootType/bootscript (0.00s)
    --- PASS: TestValidateBootType/remote (0.00s)
=== RUN   TestAccScalewayIP_importBasic
--- PASS: TestAccScalewayIP_importBasic (12.73s)
=== RUN   TestAccScalewaySecurityGroup_importBasic
--- PASS: TestAccScalewaySecurityGroup_importBasic (14.63s)
=== RUN   TestAccScalewayServer_importBasic
--- PASS: TestAccScalewayServer_importBasic (151.10s)
=== RUN   TestAccScalewayToken_importBasic
--- PASS: TestAccScalewayToken_importBasic (22.30s)
=== RUN   TestAccScalewayUserData_importBasic
--- PASS: TestAccScalewayUserData_importBasic (18.98s)
=== RUN   TestAccScalewayVolume_importBasic
--- PASS: TestAccScalewayVolume_importBasic (8.19s)
=== RUN   TestProvider
--- PASS: TestProvider (0.00s)
=== RUN   TestProvider_impl
--- PASS: TestProvider_impl (0.00s)
=== RUN   TestAccScalewayIP_Count
--- PASS: TestAccScalewayIP_Count (25.56s)
=== RUN   TestAccScalewayIP_Basic
--- PASS: TestAccScalewayIP_Basic (61.75s)
=== RUN   TestAccScalewaySecurityGroupRule_Basic
--- PASS: TestAccScalewaySecurityGroupRule_Basic (32.84s)
=== RUN   TestAccScalewaySecurityGroupRule_Count
--- PASS: TestAccScalewaySecurityGroupRule_Count (23.28s)
=== RUN   TestAccScalewaySecurityGroup_Basic
--- PASS: TestAccScalewaySecurityGroup_Basic (11.70s)
=== RUN   TestAccScalewayServer_Basic
--- PASS: TestAccScalewayServer_Basic (174.10s)
=== RUN   TestAccScalewayServer_BootType
--- PASS: TestAccScalewayServer_BootType (69.84s)
=== RUN   TestAccScalewayServer_ExistingIP
--- PASS: TestAccScalewayServer_ExistingIP (137.15s)
=== RUN   TestAccScalewayServer_Volumes
--- PASS: TestAccScalewayServer_Volumes (160.96s)
=== RUN   TestAccScalewayServer_SecurityGroup
--- PASS: TestAccScalewayServer_SecurityGroup (227.32s)
=== RUN   TestGetSSHKeyFingerprint
--- PASS: TestGetSSHKeyFingerprint (0.00s)
=== RUN   TestAccScalewaySSHKey_Basic
--- PASS: TestAccScalewaySSHKey_Basic (32.16s)
=== RUN   TestAccScalewayToken_Basic
--- PASS: TestAccScalewayToken_Basic (34.44s)
=== RUN   TestAccScalewayToken_Expiry
--- PASS: TestAccScalewayToken_Expiry (19.44s)
=== RUN   TestAccScalewayUserData_Basic
--- PASS: TestAccScalewayUserData_Basic (25.95s)
=== RUN   TestAccScalewayVolumeAttachment_Basic
--- PASS: TestAccScalewayVolumeAttachment_Basic (160.50s)
=== RUN   TestAccScalewayVolume_Basic
--- PASS: TestAccScalewayVolume_Basic (9.89s)
PASS
ok  	github.com/terraform-providers/terraform-provider-scaleway/scaleway	1493.039s
```

### CODE REVIEW

**Comment by meyskens** [comment]
:+1:

File: scaleway/resource_server.go [hunk 1]
```diff
@@ -41,6 +41,19 @@ func resourceScalewayServer() *schema.Resource {
 				Description:  "The instance type of the server",
 				ValidateFunc: validateServerType,
 			},
+			"boot_type": {
+				Type:         schema.TypeString,
+				Optional:     true,
+				ForceNew:     true,
+				Description:  "The boot_type of the server",
+				ValidateFunc: validateBootType,
+				DiffSuppressFunc: func(k, old, new string, d *schema.ResourceData) bool {
```
Comment (tombuildsstuff) [inline scaleway/resource_server.go]: Given this value would have been set to “bootscript” in the past, I think we can remove this in favour of a default value or a State migration which sets a default?

File: scaleway/resource_server.go [hunk 2]
```diff
@@ -41,6 +41,19 @@ func resourceScalewayServer() *schema.Resource {
 				Description:  "The instance type of the server",
 				ValidateFunc: validateServerType,
 			},
+			"boot_type": {
+				Type:         schema.TypeString,
+				Optional:     true,
+				ForceNew:     true,
+				Description:  "The boot_type of the server",
+				ValidateFunc: validateBootType,
```
Comment (tombuildsstuff) [inline scaleway/resource_server.go]: There’s a built-in validation function provided by Terraform Core that we can use here:

validation.StringInSlice([]string{
  “bootscript”,
  “local”,
}, true)

Where the true stands for case sensitive

File: scaleway/resource_server.go (hunk 1 above)
Comment (tombuildsstuff) [reply scaleway/resource_server.go]: Infact - is this computed?

File: scaleway/resource_server.go (hunk 2 above)
Comment (nicolai86) [reply scaleway/resource_server.go]: nice! didn't know about this.

File: scaleway/resource_server.go (hunk 1 above)
Comment (nicolai86) [reply scaleway/resource_server.go]: come to think about it, Computed and Optional I'd say. I'll adjust accordingly.

**Comment by nicolai86** [comment]
ty for the review @tombuildsstuff . I've updated the PR accordingly.

File: website/docs/r/server.html.markdown [hunk 3]
```diff
@@ -34,6 +34,7 @@ The following arguments are supported:
 * `image` - (Required) base image of server
 * `type` - (Required) type of server
 * `bootscript` - (Optional) server bootscript
+* `boot_type` - (Optional) server boot_type (`local`, `bootscript`)
```
Comment (tombuildsstuff) [inline website/docs/r/server.html.markdown]: might be worth making this `the boot mechanism for this server. Possible values include `bootscript` and `local`

**Pull request merged**